addons/mail/static/src/js/attachment_box.js

```diff
@@ -64,6 +67,68 @@ var AttachmentBox = Widget.extend({
         this.currentResModel = record.model;
     },
 
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * Calls controller to upload file, separate method is created
+     * to use it in tests to patch it.
+     *
+     * @param {string} controllerUrl
+     * @param {FormData} formData
+     */
+    _callUploadAttachment: function (controllerUrl, formData) {
+        return $.ajax({
+            url: controllerUrl,
+            type: "POST",
+            enctype: 'multipart/form-data',
+            processData: false,
+            contentType: false,
+            data: formData,
+            success: (result) => {
+                var $el = $(result);
+                $.globalEval($el.contents().text());
+            }
+        });
+    },
+    /**
+     * Making sure that dragging content is external files.
+     * Ignoring other content draging like text.
+     *
+     * @private
+     * @param {DataTransfer} dataTransfer
+     * @returns {boolean}
+     */
+    _isDragSourceExternalFile: function (dataTransfer) {
+        var DragDataType = dataTransfer.types;
+        if (DragDataType.constructor === DOMStringList) {
+            return DragDataType.contains('Files');
+        }
+        if (DragDataType.constructor === Array) {
+            return DragDataType.indexOf('Files') !== -1;
+        }
+        return false;
+    },
+    /**
+     * Upload attachment with drag & drop feature.
+     *
+     * @private
+     * @param {Array<File>} params.files
+     */
+    _processAttachmentChange: function (files) {
+        var self = this;
+        var $form = this.$('form.o_form_binary_form');
+        var formData = new FormData();
+        $form.find("input").each((index, input) => {
```

Comment:
use for-of loop instead. This works on array, so you may need to convert DomList to array using `[ ...<DomList> ]`